lualib/resty/events/broker.lua

```diff
@@ -73,19 +71,87 @@ local function broadcast_events(self, unique, data)
     log(DEBUG, "event published to ", n, " workers")
 end
 
+local function read_thread(self, worker_connection)
+    while not terminating(self, worker_connection) do
+        local data, err = worker_connection:recv_frame()
+        if err then
+            if not is_timeout(err) then
+                return nil, "failed to read event from worker: " .. err
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        if not data then
+            if not exiting() then
+                log(ERR, "did not receive event from worker")
+            end
+            goto continue
+        end
+
+        local event_data, err = decode(data)
+        if not event_data then
+            if not exiting() then
+                log(ERR, "failed to decode event data: ", err)
+            end
+            goto continue
+        end
+
+        -- unique event
+        local unique = event_data.spec.unique
+        if unique then
+            if self._uniques:get(unique) then
+                log(DEBUG, "unique event is duplicate: ", unique)
+                goto continue
+            end
+
+            self._uniques:set(unique, 1, self._opts.unique_timeout)
+        end
+
+        -- broadcast to all/unique workers
+        broadcast_events(self, unique, event_data.data)
+
+        ::continue::
+    end -- while not terminating
+
+    return true
+end
+
+local function write_thread(self, worker_connection)
+    while not terminating(self, worker_connection) do
+        local payload, err = self._clients[worker_connection]:pop()
+        if not payload then
+            if not is_timeout(err) then
+                return nil, "semaphore wait error: " .. err
+            end
+
+            goto continue
+        end
+
+        local _, err = worker_connection:send_frame(payload)
+        if err then
+            return nil, "failed to send event: " .. err
+        end
+
+        ::continue::
+    end -- while not terminating
+
+    return true
```

Comment:
```suggestion
```

Reply:
same here, it can happen when loop quits.